docs/proposals/15-manage-bastions-and-ssh-key-pair-rotation.md

```diff
@@ -0,0 +1,201 @@
+    - Controller reads `cloudprovider` credentials from seed-shoot namespace
+    - Deploy infrastructure resources
+        - Bastion VM. User data similar to https://github.com/gardener/gardenctl/blob/1e3e5fa1d5603e2161f45046ba7c6b5b4107369e/pkg/cmd/ssh.go#L160-L171. Writes `status.id_rsa.pub` into `authorized_keys` file.
+        - create security groups / firewall rules etc.
+    - Updates status of `Bastion` resource:
+        - With bastion IP under `status.bastionIP`
+        - Sets `status.state` to `Ready` on resource so that the client knows when to initiate the SSH connection
+5. `gardenlet`
+    - Once the `Bastion` resource is in ready state, it syncs back the state to the garden cluster
+6. gardenctl
+    - initiates SSH session
+        - reads `status["id_rsa.enc"]`, decrypts it with users private PGP key
+        - reads bastion IP from `status.bastionIP`
+        - reads the private key from the SSH key pair for the shoot node
+        - opens SSH to the bastion and from there to the respective shoot node
+    - runs heartbeat in parallel as long as the SSH session is open by annotating the `Bastion` resource with `operations.gardener.cloud/operation: keepalive`
+7. `GCM`:
+    - Once `status.expirationDate` is reached, the `Bastion` will be marked for deletion
+8. `gardenlet`:
+    - Once the `Bastion` resource in the garden cluster is marked for deletion, it marks the `Bastion` resource in the seed for deletion
+9. Gardener extension provider <infra> / Bastion Controller on Seed:
+    - all created resources will be cleaned up
+    - On succes, removes finalizer on `Bastion` resource in seed
+10. `gardenlet`:
+    - removes finalizer on `Bastion` resource in garden cluster
+
+### Resource Example
+
+`Bastion` resource in the garden cluster
+```yaml
+apiVersion: operations.gardener.cloud/v1alpha1
+kind: Bastion
+metadata:
+  generateName: cli-
+  name: cli-abcdef
+  namespace: garden-myproject
+  annotations:
+    operations.gardener.cloud/created-by: foo # set by the mutating webhook
+    operations.gardener.cloud/last-heartbeat-at: "2021-03-19T11:58:00Z"
+    # operations.gardener.cloud/operation: keepalive # this annotation is removed by the mutating webhook and the last-heartbeat timestamp and/or the status.expirationDate will be updated accordingly
+spec:
+  shootRef: # namespace cannot be set / it's the same as .metadata.namespace
+    name: my-cluster
+
+  # seedName: aws-eu2 # is set by the mutating webhook
+  # providerType: aws # is set by the mutating webhook
+
+  publicKey: LS0tLS1CRUdJTiBQR1AgUFVCTElDIEtFWSBCTE9DSy0tLS0tCi4uLgotLS0tLUVORCBQR1AgUFVCTElDIEtFWSBCTE9DSy0tLS0tCg== # user's PGP public key.
+
+  clientIP: # external ip of the user
+    ipv4: 1.2.3.4
+    # ipv6: ::1
+
+status:
+  # the following fields are managed by the controller in the seed and synced by gardenlet
+  bastionIP: 1.2.3.5
+  state: Ready
```

Comment:
What about 
```suggestion
  phase: Ready
```

Or, why not have a `.status.lastOperation` here similar to other resources?
Or, why not only work with conditions?

Reply:
I talked with @xrstf and we agreed to go for conditions instead of a `state` or `lastOperation`.
The client will then look out for a `BastionReady` condition. If true, it will assume everything is ready and initiate the ssh connection. WDYT?

Reply:
Sure, sounds good, thanks